system-addon/lib/DownloadsManager.jsm

```diff
@@ -0,0 +1,154 @@
+      hostname: new URL(downloadedItem.source.url).hostname,
+      url: downloadedItem.source.url,
+      path: downloadedItem.target.path,
+      title: element.displayName,
+      description: `${element.sizeStrings.stateLabel} \u2014 ${element.fileType}`,
+      referrer: downloadedItem.source.referrer,
+      date_added: downloadedItem.endTime
+    };
+  }
+
+  init(store) {
+    this._store = store;
+    this._browser = Services.appShell.hiddenDOMWindow;
+    this._downloadData = DownloadsCommon.getData(this._browser.ownerGlobal, true, false, true);
+    this._downloadData.addView(this);
+  }
+
+  onDownloadAdded(download) {
+    const elem = new DownloadElement(download, this._browser);
+    const downloadedItem = elem.download;
+    if (!this._viewableDownloadItems.has(downloadedItem.source.url)) {
+      this._viewableDownloadItems.set(downloadedItem.source.url, elem);
+      this._store.dispatch({type: at.DOWNLOAD_CHANGED});
+    }
+  }
+
+  onDownloadRemoved(download) {
+    if (this._viewableDownloadItems.has(download.source.url)) {
+      this._viewableDownloadItems.delete(download.source.url);
+      this._store.dispatch({type: at.DOWNLOAD_CHANGED});
+    }
+  }
+
+  get downloads() {
+    let results = [];
+    for (const url of this._viewableDownloadItems.keys()) {
+      const elem = this._viewableDownloadItems.get(url);
+      // We only want downloads that are less than 36 hours old
+      if (elem.downloadedSince < THIRTY_SIX_HOURS) {
+        const formattedDownloadForHighlights = this.formatDownload(elem);
+        results.push(formattedDownloadForHighlights);
+      }
+    }
+    return results.sort((a, b) => a.date_added < b.date_added);
+  }
+
+  openFile(url) {
+    const download = this.findDownloadedItem(url);
+    download.downloadsCmd_open();
+  }
+
+  showFile(url) {
+    const download = this.findDownloadedItem(url);
+    download.downloadsCmd_show();
+  }
+
+  copyFile(url) {
+    const download = this.findDownloadedItem(url);
+    download.downloadsCmd_copyLocation();
+  }
+
+  goToDownloadPage(url) {
+    const download = this.findDownloadedItem(url);
+    download.downloadsCmd_openReferrer();
+  }
+
+  removeDownload(url) {
+    const download = this.findDownloadedItem(url);
+    download.downloadsCmd_delete();
+  }
+
+  findDownloadedItem(url) {
+    return this._viewableDownloadItems.get(url);
+  }
+
+  uninit() {
+    if (this._downloadData) {
+      this._downloadData.removeView(this);
+    }
+  }
+
+  onAction(action) {
+    switch (action.type) {
+      case at.COPY_DOWNLOAD_LINK:
+        this.copyFile(action.data.url);
+        break;
+      case at.GO_TO_DOWNLOAD_PAGE:
+        this.goToDownloadPage(action.data.url);
+        break;
+      case at.REMOVE_DOWNLOAD_FILE:
+        this.removeDownload(action.data.url);
+        break;
+      case at.SHOW_DOWNLOAD_FILE:
+        this.showFile(action.data.url);
+        break;
+      case at.OPEN_DOWNLOAD_FILE:
+        this.openFile(action.data.url);
```

Comment:
Eh.. these 5 actions could just just do something like
```js
  case OPEN:
    downloadsCmd = "open";
    break;
…
if (downloadsCmd) {
  this.findDownloadedItem(action.data.url)[`downloadsCmd_${downloadsCmd}`]();
}
```

Too clever? ;)

Reply:
Yeah I thought about this too. I don't mind being too clever ;)